File: backend/app/schemas/crop.py
from datetime import date, datetime

from pydantic import BaseModel

from app.schemas.common import ORMModel


class CropBase(BaseModel):
    farm_id: int
    crop_type: str
    crop_variety: str | None = None
    sowing_date: date | None = None
    growth_stage: str | None = None
    irrigation_type: str | None = None
    field_size: str | None = None
    notes: str | None = None


class CropCreate(CropBase):
    pass


class CropUpdate(BaseModel):
    crop_type: str | None = None
    crop_variety: str | None = None
    sowing_date: date | None = None
    growth_stage: str | None = None
    irrigation_type: str | None = None
    field_size: str | None = None
    notes: str | None = None


class CropRead(CropBase, ORMModel):
    id: int
    created_at: datetime
    updated_at: datetime

File: frontend/lib/models/crop.dart
class AppCrop {
  final int id;
  final int farmId;
  final String cropType;
  final String? cropVariety;
  final String? sowingDate;
  final String? growthStage;
  final String? irrigationType;
  final String? fieldSize;
  final String? notes;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  AppCrop({
    required this.id,
    required this.farmId,
    required this.cropType,
    this.cropVariety,
    this.sowingDate,
    this.growthStage,
    this.irrigationType,
    this.fieldSize,
    this.notes,
    this.createdAt,
    this.updatedAt,
  });

  factory AppCrop.fromJson(Map<String, dynamic> json) {
    return AppCrop(
      id: json['id'] as int,
      farmId: json['farm_id'] as int,
      cropType: json['crop_type'] as String,
      cropVariety: json['crop_variety'] as String?,
      sowingDate: json['sowing_date'] as String?,
      growthStage: json['growth_stage'] as String?,
      irrigationType: json['irrigation_type'] as String?,
      fieldSize: json['field_size'] as String?,
      notes: json['notes'] as String?,
      createdAt: json['created_at'] != null ? DateTime.tryParse(json['created_at']) : null,
      updatedAt: json['updated_at'] != null ? DateTime.tryParse(json['updated_at']) : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'farm_id': farmId,
        'crop_type': cropType,
        'crop_variety': cropVariety,
        'sowing_date': sowingDate,
        'growth_stage': growthStage,
        'irrigation_type': irrigationType,
        'field_size': fieldSize,
        'notes': notes,
      };
}
